python/dgl/data/utils.py

```diff
@@ -482,3 +484,198 @@ def add_nodepred_split(dataset, ratio, ntype=None):
+        "in_test_mask",
+        "out_valid_mask",
+        "out_test_mask",
+    ]
+    split_masks = {}
+
+    for mask_name, node_indices in zip(mask_names, node_split):
+        split_mask = idx2mask(node_indices, num_nodes)
+        split_masks[mask_name] = generate_mask_tensor(split_mask)
+
+    return split_masks
+
+
+def add_node_property_split(
+    dataset, part_ratios, property_name, ascending=True, random_seed=None
+):
+    """Create a data split with a distributional shift based on some node property.
+
+    It splits each graph in the given dataset into training, ID and OOD validation, ID and OOD
+    testing parts for transductive node prediction task with structural distributional shifts.
+    As a result, it creates 5 associated node mask arrays for each graph:
+        - 3 for the ID nodes: ``'in_train_mask'``, ``'in_valid_mask'``, ``'in_test_mask'``,
+        - and 2 for the OOD nodes: ``'out_valid_mask'``, ``'out_test_mask'``.
+
+    Following `Evaluating Robustness and Uncertainty of Graph Models Under Structural
+    Distributional Shifts <https://arxiv.org/abs/2302.13875v1>`__, this function implements
+    3 particular strategies for inducing distributional shifts in graph — based on
+    **popularity**, **locality** or **density**.
+
+    Parameters
+    ----------
+    dataset : DGLDataset or list of :class:`DGLGraph`
+        The dataset to induce structural distributional shift in.
+    part_ratios : list
+        A list of 5 ratio values for training, ID validation, ID test,
+        OOD validation and OOD test parts. The values must sum to 1.0.
+    property_name : str
+        The node property name to split the dataset by. Must be
+        ``'popularity'``, ``'locality'`` or ``'density'``.
+    ascending : bool, optional
+        Whether to sort nodes in the ascending order of a particular
+        node property, so that more shifted OOD nodes have greater values
+        of the computed property (default: True)
+    random_seed : int, optional
+        Random seed to fix for the initial permutation of nodes. It is
+        used to create a random order for the nodes that have the same
+        property values. (default: None)
+
+    Examples
+    --------
+    >>> dataset = dgl.data.AmazonCoBuyComputerDataset()
+    >>> print('in_valid_mask' in dataset[0].ndata)
+    False
+    >>> part_sizes = [0.3, 0.1, 0.1, 0.3, 0.2]
+    >>> property_name = 'popularity'
+    >>> dgl.data.utils.add_node_property_split(dataset, part_sizes, property_name)
+    >>> print('in_valid_mask' in dataset[0].ndata)
+    True
+    """
+
+    assert property_name in [
+        "popularity",
+        "locality",
+        "density",
+    ], "The name of property has to be 'popularity', 'locality', or 'density'"
+
+    assert (
+        len(part_ratios) == 5
+    ), "The list of part ratios must contain 5 values"
+
+    try:
+        import graph_tool as gt
+        from graph_tool import centrality, clustering, generation
```

Comment:
This introduces new dependencies. Is it possible to use existing DGL dependency like NetworkX?

Reply:
I have reimplemented the whole introduced functional using `networkx` and removed everything related to `graph-tool`.